parser/impl/parser.cpp

```diff
@@ -469,17 +469,201 @@ sPtr<ast::IfStatement> Parser::parseIfStatement() {
 // ---- @aabounegm
 
 sPtr<ast::Expression> Parser::parseExpression() {
-    // TODO: Concrete implementation
+    return parseBinaryExpression(0);
+}
+
+int Parser::opPrec(TokenType token) {
+    switch (token) {
+    case TokenType::Or:  // or
+    case TokenType::Xor: // xor
+        return 1;
+    case TokenType::And: // and
+        return 2;
+    case TokenType::Eq:      // =
+    case TokenType::Neq:     // /=
+    case TokenType::Less:    // <
+    case TokenType::Leq:     // <=
+    case TokenType::Greater: // >
+    case TokenType::Geq:     // >=
+        return 3;
+    case TokenType::Add: // +
+    case TokenType::Sub: // -
+        return 4;
+    case TokenType::Mul: // *
+    case TokenType::Div: // /
+    case TokenType::Mod: // %
+        return 5;
+    case TokenType::Not:       // not
+    case TokenType::OpenParen: // ()
+        return 6;
+    case TokenType::Dot:       // .
+    case TokenType::OpenBrack: // []
+        return 7;
+    default:
+        return -1; // unknown op
+    }
+}
+
+bool Parser::isPrimitive(TokenType token) {
+    return token == TokenType::Identifier || token == TokenType::Int ||
+           token == TokenType::Real || token == TokenType::True ||
+           token == TokenType::False;
+}
+
+sPtr<ast::Expression> Parser::parseUnaryExpression() {
+    Token tok = m_lexer.Peek();
+    ast::Expression expr;
+    if (opPrec(tok.type) >= 0) {
+        if (tok.type == TokenType::Not || tok.type == TokenType::Sub ||
+            tok.type == TokenType::Add) {
+
+            tok = m_lexer.Next();
+            expr.operation = tok;
+            expr.operand1 = parseUnaryExpression();
+            return std::make_shared<ast::Expression>(expr);
+
+        } else if (tok.type == TokenType::OpenParen) {
+            tok = m_lexer.Next();
+            sPtr<ast::Expression> expr = parseBinaryExpression(0);
+            if (m_lexer.Peek().type == TokenType::CloseParen) {
+                tok = m_lexer.Next();
+                return expr;
+            }
+            // error
+            m_errors.push_back(Error{
+                .pos = tok.pos,
+                .message = "Expected to find ')'",
+            });
+            return nullptr;
+        } else {
+            // throw error
+            m_errors.push_back(Error{
+                .pos = tok.pos,
+                .message = "Expected to find unary operator",
+            });
+            return nullptr;
+        }
+    } else if (isPrimitive(tok.type)) {
+        tok = m_lexer.Next();
+        // check if More Then Identifier
```

Comment:
What does that mean? How can there be multiple consecutive identifiers without any operator in between?

Reply:
1. Function call - 100% hasn't operator `a:=b(3, 5, c)`. Moreover, it should be distinguished with parenthesis inside an arithmetical expression.
2. Array index call - he problem is that `[3] ` still isn't unary or binary operation by its meaning.
3. Dot notation - it's mandatory to check that it is not sequence of separated expressions, but it is a sequence of separated `Identifiers`. 

Moreover, we can have a sequence like this:

`a.b[7].n`

But not like:

`a.b(2)` or `c[7](1)`

I should also be controlled during the parsing stage.

Reply:
1. Still don't understand how that's related to what's written in the code
2. Array element access **is** a Binary expression, just like the dot-access. It's first operand is the array and the second one is the expression representing the index
3. right, but to make it easier in this stage, we can apply what I suggested yesterday about making this tree simple and then performing more checks later with different passes on the AST. Moreover, if we allow `a.b(2)` or `c[7](1)` at this stage and check it at a later one, we make future changes easier (for example, introducing an array of functions or classes). I am not saying we _have to_ care about that at all in this project, but that it is good practice and would teach us better software engineering

P.S.: The tone may not be clear from the text, but read it as more of a friendly discussion and not criticism 😁